config/configmapprovider/composite.go

```diff
@@ -0,0 +1,294 @@
+	ctx context.Context,
+	composeFunc configComposeFunc,
+	subRetrieved retrievedList,
+) (*compositeRetrieved, error) {
+	// Unfortunately Retrieved.Get() cannot return an error, but it is
+	// possible that composeFunc returns an error. Because of that we cannot
+	// do the composition using composeFunc in our Get() implementation and
+	// have to do it here so that we can return an error.
+	// TODO: change Retrieved.Get() to allow returning an error and move
+	// the following code to our Get().
+
+	// Call Get() on all sub-retrieved and combine the results using the composeFunc.
+	composedMap, err := subRetrieved.getAllAndCompose(composeFunc)
+	if err != nil {
+		// Didn't work. Close every WatchableRetrieved that may be on our list.
+		for _, retr := range subRetrieved {
+			if watchable, ok := retr.(WatchableRetrieved); ok {
+				err2 := watchable.Close(ctx)
+				err = multierr.Append(err, err2)
+			}
+		}
+		return nil, err
+	}
+
+	// Create a new cancellable context for long-running operations.
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	m := &compositeRetrieved{
+		composedMap:        composedMap,
+		composeFunc:        composeFunc,
+		subRetrieved:       subRetrieved,
+		watchContext:       ctx,
+		cancelWatchContext: cancelFunc,
+	}
+
+	return m, nil
+}
+
+func (cr *compositeRetrieved) Get() *config.Map {
+	// WatchableRetrieved requires that Get() is never called concurrently
+	// with WatchForUpdate() which means a race on cr.composedMap is not possible
+	// since the only place where we modify it is in WatchForUpdate(), so there
+	// is no need to protect cr.composedMap from races.
+	return cr.composedMap
+}
+
+func (cr *compositeRetrieved) WatchForUpdate() error {
+	// Create one go routine per sub-retrieved and start waiting for WatchForUpdate()
+	// of any Watchable sub-retrieved to return.
+
+	// Channel used to signal that sub-retrieved WaitForUpdate() returned.
+	readyWatchableCh := make(chan readyWatchable)
+
+	// Count how many of the sub-retrieved are watchable.
+	watchableCount := 0
+	for _, subRetrieved := range cr.subRetrieved {
+		if watchable, ok := subRetrieved.(WatchableRetrieved); ok {
+			// This is a Watchable sub-retrieved.
+			// Start watching for updates from this WatchableRetrieved.
+			watchableCount++
+			go func() {
+				err := watchable.WatchForUpdate()
+				select {
+				case readyWatchableCh <- readyWatchable{watchable, err}:
+					// We have an update, signal it to the outer func.
+				case <-cr.watchContext.Done():
+					// We are instructed to terminate. Just return.
+				}
+			}()
+		}
+	}
+
+	select {
+	case readyWatchable := <-readyWatchableCh:
+
+		// At least one watcher fired. Close all sub-Retrieved. We don't need
+		// to close the one that fired the signal (readyWatchable.watchable),
+		// so provide it as an exception to closeAllExcept().
+		if err := cr.closeAllExcept(cr.watchContext, readyWatchable.watchable); err != nil {
+			return err
+		}
```

Comment:
Not sure why you don't close that as well. I am not sure that we want the limitation to be when `WatchForUpdate` returns to also close, maybe we want to `WatchForUpdate` again? If this is the case you need to document that when returns Close is not needed to be called.